### ISSUE TITLE
[FormRequest] validated() method causes the validation to be called again.

### ISSUE THREAD

**Issue by DmitryNiko** [laravel/framework#25736]
- Laravel Version: 5.7.5
- PHP Version: 7.2.9
- Database Driver & Version: MySQL 5.7.23

### Description:
The problem is getting the retrieve the validated input data with method `validated()`. When it is called, it re-initiates the request validation. It seems like 941c8c7be2ad4441cdba12053c15f5dc46980707 commit breaks this.

### Steps To Reproduce:
1. Controller:
```
use App\Http\Requests\ClientStoreRequest;

class ClientController extends Controller
{
    public function store(ClientStoreRequest $request)
    {
        $this->clientService->createClient($request->validated());

        return redirect()->route('app.clients')->with('status', 'Client was successfully created!');
    }
}
```
2. FormRequest:
```
use Illuminate\Foundation\Http\FormRequest;
use Illuminate\Validation\Rule;

class ClientStoreRequest extends FormRequest
{
    public function rules()
    {
        return [
	    'full_name' => 'required|string|max:50',
	    'phone' => [
	        'required',
	        'max:50',
	        'regex:/^(\+[0-9]+)$/',
	        Rule::unique('clients')->where(function($query){
	            return $query->where('account_id', auth()->user()->account_id);
	        })
	    ],
	    'email' => 'nullable|email|max:50',
	    'info' => 'max:200'
        ];
    }
}
```
3. Debugbar:
Since the rules have a check in the database, we can see a duplicated query to the database, which means that the method rules() was called twice.
![debugbar_546654652654](https://user-images.githubusercontent.com/25366375/45906505-1286cd80-bdfd-11e8-9ccd-a6ddf56f6c10.jpg)

**Comment by staudenmeir** [comment]
Please take a look at this @ttsuru. After #25128, calling `validated()` unnecessarily re-validates the data.

Can we store the validated data in `ValidatesWhenResolvedTrait::validateResolved()` and then just return it when `validated()` is called?

**Comment by ttsuru** [comment]
1 is breaking change method to `Validator::validate()`. `validate()` validate only once.
2 requires to add contracts `Illuminate\Contracts\Validation\Validator::validated()`
3 requires to add contracts `Illuminate\Contracts\Validation\Validator::getRules()`

**Comment by gofish543** [comment]
@ankur91 Sorry about the duplication

I want to add that the scope of this issue is not limited to a few database queries being done twice. While annoying and inefficient, it isn't a breaking fault.

However, with regards to validation that uses One Time Password (OTP) keys or passwords, this issue becomes more of a breaking fault rather than an inconvenience. 

An example of this is Google Recaptcha. I'm going to have to downgrade my Laravel version, to before the breaking commit because the other option of disabling recaptcha is just not an option.

**Comment by mohammedmanssour** [comment]
Hi,
@ttsuru  I think you're working to resolve this, right? If yes, where are you at it?
for me, adding the `Illuminate\Contracts\Validation\Validator::validated()` is the best choice we have. but I don't seem to understand why It's not a recommended option?
if you're not working on this one, please let me know, because I'd like to get this as my first contribution to Laravel 😃

**Comment by DmitryNiko** [comment]
@mohammedmanssour  For me, this error is still a big problem in my app, which does not allow me upgrade to the latest version of Laravel. I also hope that this will be fixed soon.

**Comment by ttsuru** [comment]
pull request #26417.
Because no one was response, I hesitated pull request.

**Comment by driesvints** [comment]
Will be fixed once 5.8 is released.

**Comment by ttsuru** [comment]
@Rux77 Add new test #26484
I add test. Please check this issue is fixed.

**Comment by driesvints** [comment]
@ttsuru can you confirm that this is actually fixed?

**Comment by Rux77** [comment]
This problem is a big issue for a project I am working on..
For now, to solve this.. I am using this method ( which was in **laravel 5.6.*** ) and removed in later releases: 
```
    /**
     * Get the validated data from the request.
     *
     * @return array
     */
    public function validated()
    {
        $rules = $this->container->call([$this, 'rules']);
        return $this->only(collect($rules)->keys()->map(function ($rule) {
            return explode('.', $rule)[0];
        })->unique()->toArray());
    }
```

**Comment by driesvints** [comment]
Please see my comment on said PR.

**Comment by gofish543** [comment]
Can this be merged into 5.7?
This was a critical fault introduced when 5.7 was released and it sounds like I am not the only one who is forced to use 5.6 until this is resolved.

**Comment by driesvints** [comment]
The current solution cant be merged in 5.7 because it contains breaking changes.

**Comment by LastDragon-ru** [comment]
> The current solution cant be merged in 5.7 because it contains breaking changes.

It is not a "breaking changes" it is a "Changed realization" https://github.com/laravel/framework/pull/26314 😄

**Comment by robjbrain** [comment]
Adding my voice to this completely breaking form requests, as others have said if you validate via a 3rd party api a double validation won't work. As well as any other expensive validation from database queries or custom methods.

If you are really serious about waiting till 5.8 (3 months?) the you should update the docs to warn people not to use the example if they're using expensive api calls, databases queries or custom methods.

**Comment by driesvints** [comment]
We're probably not going to fix this in the current 5.7 release. I realize that the current implementation isn't ideal but it's also not a critical bug. This will be fixed in the 5.8 release.

**Comment by robjbrain** [comment]
The easiest temporary fix for anyone having this issue would be to extend the FormRequest class and intercept the validated() method e.g:

```php
namespace App\Http\Requests;

use \Illuminate\Validation\ValidationException;
use Illuminate\Foundation\Http\FormRequest as LaravelFormRequest;

class FormRequest extends LaravelFormRequest
{
    /**
     * The validated data
     *
     * @var array
     */
    protected $validatedData;

    /**
     * Return the validated data.
     *
     * @return array
     *
     * @throws \Illuminate\Validation\ValidationException
     */
    public function validated()
    {
        if ($this->fails()) {
            throw new ValidationException($this);
        }

        return $this->validatedData ?: $this->validatedData = parent::validated();
    }
}
```

It's not how it should be fixed within Laravel obviously, but it's the easiest fix and should be compatible with any changes made between now and 5.8.

**Comment by taylorotwell** [comment]
I think we should fix this in the 5.7 release and not change the contract.

**Comment by henriquerdd** [comment]
Is this really fixed? I'm using laravel 5.8 and during a http test it happened to me.
I need to do some api calls on my validation (2 to be precise), so i've mocked my api clients:

```
$response = \Mockery::mock(\Psr\Http\Message\ResponseInterface::class);

$response->shouldReceive('getStatusCode')->andReturn(200);

$this->mock(ProductsApiClient::class, function ($mock) use ($response) {
    $mock->shouldReceive('head')->once()->andReturn($response);
});

$this->mock(CollectionsApiClient::class, function ($mock) use ($response) {
    $mock->shouldReceive('head')->once()->andReturn($response);
});
```

The validation logic is as follow:

```
public function rules()
{
    return [
        'productUuid' => ['required', 'uuid', function ($attribute, $value, $fail) {

            if (!$this->productExists($value)) {
                return $fail("Product '$value' not found");
            }
        }],
        'collectionUuid' => ['required', 'uuid', function ($attribute, $value, $fail) {

            if (!$this->collectionExists($value)) {
                return $fail("Collection '$value' not found");
            }
        }]
    ];
}

private function productExists($productUuid)
{
    $response = $this->productsApiClient->head("/products/$productUuid");
    return $response->getStatusCode() == 200;
}

private function collectionExists($collectionUuid)
{
    $response = $this->collectionsApiClient->head("/collections/$collectionUuid");
    return $response->getStatusCode() == 200;
}
```

The test itself:

```

$response = $this->json('POST', '/something', $body);

$response
    ->assertStatus(201)
    ...
```

But when i run the tests, phpunit screams:

1) Tests\Feature\SomeTest::testCreationWithValidInput
Mockery\Exception\InvalidCountException: Method head(<Any Arguments>) from Mockery_3_App_ProductsApiClient should be called
 exactly 1 times but called 2 times.
